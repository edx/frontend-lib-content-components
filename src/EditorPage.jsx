import React, { useRef, useEffect, useState } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { FormattedMessage } from '@edx/frontend-platform/i18n';
import { AppProvider } from '@edx/frontend-platform/react';

import { blockTypes } from 'data/constants/app';
import { actions } from 'data/redux';

import store from 'data/store';

import TextEditor from './editors/TextEditor/TextEditor';
import VideoEditor from './editors/VideoEditor/VideoEditor';
import ProblemEditor from './editors/ProblemEditor/ProblemEditor';
import EditorFooter from './components/EditorFooter';
import EditorHeader from './components/EditorHeader';

export const messages = {
  id: 'authoring.editorpage.selecteditor.error',
  defaultMessage: 'Error: Could Not find Editor',
  description: 'Error Message Dispayed When An unsopported Editor is desired in V2',
};

export const supportedEditors = {
  [blockTypes.html]: TextEditor,
  [blockTypes.video]: VideoEditor,
  [blockTypes.problem]: ProblemEditor,
};

export const initializeApp = ({ initialize, data }) => useEffect(() => initialize(data), []);
export const prepareEditorRef = () => {
  const editorRef = useRef(null);
  const setEditorRef = (ref) => {
    editorRef.current = ref;
  };
  const [refReady, setRefReady] = useState(false);
  useEffect(() => setRefReady(true), []);
  return { editorRef, refReady, setEditorRef };
};

export const EditorPage = ({
  courseId,
  blockType,
  blockId,
  studioEndpointUrl,
  // redux
  initialize,
}) => {
  module.initializeApp({
    initialize,
    data: {
      blockId,
      blockType,
      courseId,
      studioEndpointUrl,
    },
  });

  const {
    editorRef,
    refReady,
    setEditorRef,
  } = module.prepareEditorRef();

  const Editor = supportedEditors[blockType];

  return (
    <AppProvider store={store}>
      <div className="d-flex flex-column vh-100">
        <div
          className="pgn__modal-fullscreen"
          role="dialog"
          aria-label={blockType}
        >
          <EditorHeader />

          {refReady && (
            (Editor !== undefined)
              ? <Editor {...{ editorRef, setEditorRef }} />
              : <FormattedMessage {...messages.couldNotFindEditor} />
          )}

          <EditorFooter />
        </div>
      </div>
    </AppProvider>
  );
};
EditorPage.defaultProps = {
  courseId: null,
  blockId: null,
  studioEndpointUrl: null,
};

EditorPage.propTypes = {
  courseId: PropTypes.string,
  blockType: PropTypes.string.isRequired,
  blockId: PropTypes.string,
  studioEndpointUrl: PropTypes.string,
  // redux
  initialize: PropTypes.func.isRequired,
};
export const mapStateToProps = () => ({});
export const mapDispatchToProps = {
  initialize: actions.app.initialize,
};

export default connect(mapStateToProps, mapDispatchToProps)(EditorPage);
