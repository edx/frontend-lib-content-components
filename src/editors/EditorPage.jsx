import React from 'react';
import PropTypes from 'prop-types';
import { Provider } from 'react-redux';

import store from './data/store';
import Editor from './Editor';

export const EditorPage = ({
  courseId,
  blockType,
  blockId,
  studioEndpointUrl,
}) => (
  <Provider store={store}>
    <Editor
      {...{
        courseId,
        blockType,
        blockId,
        studioEndpointUrl,
      }}
    />
  </Provider>
);
EditorPage.defaultProps = {
  courseId: null,
  blockId: null,
  studioEndpointUrl: null,
};

EditorPage.propTypes = {
  courseId: PropTypes.string,
  blockType: PropTypes.string.isRequired,
  blockId: PropTypes.string,
  studioEndpointUrl: PropTypes.string,
};

export default EditorPage;
