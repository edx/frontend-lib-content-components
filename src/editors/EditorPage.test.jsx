import React from 'react';
import { render, screen } from '@testing-library/react';
import EditorPage from './EditorPage';

test('rendering correctly with expected Input', () => {
  const courseId = 'course-v1:edX+DemoX+Demo_Course';
  const blockType = 'html';
  const blockId = 'block-v1:edX+DemoX+Demo_Course+type@html+block@030e35c4756a4ddc8d40b95fbbfff4d4';
  const studioEndpointUrl = 'fakeurl.com';
  render(<EditorPage
    courseId={courseId}
    blockType={blockType}
    blockId={blockId}
    studioEndpointUrl={studioEndpointUrl}
  />);
  expect(screen.getByText('Text')).toBeTruthy();
  expect(screen.getByText('Cancel')).toBeTruthy();
  expect(screen.getAllByLabelText('Close')).toBeTruthy();
  expect(screen.getByText('Add To Course')).toBeTruthy();
  expect(screen.getByText('Error: Could Not Load Text Content')).toBeTruthy();
});

test('rendering correctly with expected Error', () => {
  const courseId = 'course-v1:edX+DemoX+Demo_Course';
  const blockType = 'Smelly Garbage Xblock';
  const blockId = 'BadGarbadioU@Garbagefha4521ea ';
  const studioEndpointUrl = 'fakeurl.com';
  render(<EditorPage
    courseId={courseId}
    blockType={blockType}
    blockId={blockId}
    studioEndpointUrl={studioEndpointUrl}
  />);
  expect(screen.getByText(blockType)).toBeTruthy();
  expect(screen.getByText('Cancel')).toBeTruthy();
  expect(screen.getAllByLabelText('Close')).toBeTruthy();
  expect(screen.getByText('Add To Course')).toBeTruthy();
  expect(screen.getByText('Error: Could Not find Editor')).toBeTruthy();
});
