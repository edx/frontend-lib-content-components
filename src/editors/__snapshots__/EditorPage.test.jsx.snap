// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor Page snapshots rendering correctly with expected Input 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(Symbol.observable): [Function],
    }
  }
>
  <Editor
    blockId="block-v1:edX+DemoX+Demo_Course+type@html+block@030e35c4756a4ddc8d40b95fbbfff4d4"
    blockType="html"
    courseId="course-v1:edX+DemoX+Demo_Course"
    studioEndpointUrl="fakeurl.com"
  />
</Provider>
`;
